cli/file_fetcher.rs

```diff
@@ -2,677 +2,499 @@
-use deno_core::url::Url;
 use deno_core::ModuleSpecifier;
 use deno_fetch::reqwest;
-use log::info;
 use std::collections::HashMap;
 use std::fs;
 use std::future::Future;
 use std::io::Read;
-use std::path::Path;
 use std::path::PathBuf;
 use std::pin::Pin;
-use std::result::Result;
-use std::str;
 use std::sync::Arc;
 use std::sync::Mutex;
 
-/// Structure representing local or remote file.
-#[derive(Debug, Clone)]
-pub struct SourceFile {
-  pub url: Url,
-  pub filename: PathBuf,
-  pub types_header: Option<String>,
+const SUPPORTED_SCHEMES: [&str; 3] = ["http", "https", "file"];
+
+/// A structure representing a source file.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct File {
+  /// The path to the local version of the source file.  For local files this
+  /// will be the direct path to that file.  For remote files, it will be the
+  /// path to the file in the HTTP cache.
+  pub local: PathBuf,
+  /// For remote files, if there was an `X-TypeScript-Type` header, the parsed
+  /// out value of that header.
+  pub maybe_types: Option<String>,
+  /// The resolved media type for the file.
   pub media_type: MediaType,
-  pub source_code: String,
+  /// The source of the file as a string.
+  pub source: String,
+  /// The _final_ specifier for the file.  The requested specifier and the final
+  /// specifier maybe different for remote files that have been redirected.
+  pub specifier: ModuleSpecifier,
 }
 
 /// Simple struct implementing in-process caching to prevent multiple
 /// fs reads/net fetches for same file.
 #[derive(Clone, Default)]
-pub struct SourceFileCache(Arc<Mutex<HashMap<String, SourceFile>>>);
+struct FileCache(Arc<Mutex<HashMap<ModuleSpecifier, File>>>);
 
-impl SourceFileCache {
-  pub fn set(&self, key: String, source_file: SourceFile) {
-    let mut c = self.0.lock().unwrap();
-    c.insert(key, source_file);
+impl FileCache {
+  pub fn get(&self, specifier: &ModuleSpecifier) -> Option<File> {
+    let cache = self.0.lock().unwrap();
+    if let Some(file) = cache.get(specifier) {
+      Some(file.clone())
+    } else {
+      None
+    }
   }
 
-  pub fn get(&self, key: String) -> Option<SourceFile> {
-    let c = self.0.lock().unwrap();
-    match c.get(&key) {
-      Some(source_file) => Some(source_file.clone()),
-      None => None,
-    }
+  pub fn insert(&self, specifier: ModuleSpecifier, file: File) -> Option<File> {
+    let mut cache = self.0.lock().unwrap();
+    cache.insert(specifier, file)
   }
 }
 
-const SUPPORTED_URL_SCHEMES: [&str; 3] = ["http", "https", "file"];
-
-#[derive(Clone)]
-pub struct SourceFileFetcher {
-  source_file_cache: SourceFileCache,
-  cache_blocklist: Vec<String>,
-  use_disk_cache: bool,
-  no_remote: bool,
-  cached_only: bool,
-  http_client: reqwest::Client,
-  // This field is public only to expose it's location
-  pub http_cache: HttpCache,
+/// Indicates how cached source files should be handled.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub enum Cache {
```

Comment:
Bike-shed: `CacheConfig`?

Reply:
If `Cache` is confusing, I would prefer `CacheSetting` versus `CacheConfig`.